File: python/tests/test_food.py
import unittest

from snake_game.food import FoodManager


class SmallestCellRng:
    """fake random.choice, always picks smallest cell so spawns are predictable"""

    def choice(self, seq):
        return sorted(seq)[0]


class FoodManagerTests(unittest.TestCase):
    def test_reset_spawns_the_configured_count(self):
        food = FoodManager(cols=3, rows=3, count=2, rng=SmallestCellRng())
        food.reset(occupied={(0, 0)})
        self.assertEqual(len(food.items), 2)

    def test_reset_never_spawns_on_an_occupied_cell(self):
        food = FoodManager(cols=3, rows=3, count=2, rng=SmallestCellRng())
        food.reset(occupied={(0, 0)})
        self.assertNotIn((0, 0), food.items)

    def test_consume_removes_the_eaten_item_and_respawns_a_new_one(self):
        food = FoodManager(cols=3, rows=3, count=1, rng=SmallestCellRng())
        food.reset(occupied=set())
        eaten = food.items[0]

        consumed = food.consume(eaten, occupied={eaten})

        self.assertTrue(consumed)
        self.assertEqual(len(food.items), 1)
        self.assertNotIn(eaten, food.items)

    def test_consume_ignores_a_cell_with_no_food(self):
        food = FoodManager(cols=3, rows=3, count=1, rng=SmallestCellRng())
        food.reset(occupied=set())
        self.assertFalse(food.consume((2, 2), occupied=set()))


if __name__ == '__main__':
    unittest.main()

File: cpp/tests/test_food.cpp
#include <algorithm>

#include "doctest.h"
#include "food.hpp"

TEST_CASE("reset spawns the configured count") {
    FoodManager food(3, 3, 2, /*seed=*/42);
    food.Reset({Cell{0, 0}});
    CHECK(food.Items().size() == 2);
}

TEST_CASE("reset never spawns on an occupied cell") {
    FoodManager food(3, 3, 2, /*seed=*/42);
    food.Reset({Cell{0, 0}});
    bool onOccupied = std::find(food.Items().begin(), food.Items().end(), Cell{0, 0}) != food.Items().end();
    CHECK_FALSE(onOccupied);
}

TEST_CASE("consume removes the eaten item and respawns a new one") {
    FoodManager food(3, 3, 1, /*seed=*/42);
    food.Reset({});
    Cell eaten = food.Items().front();

    bool consumed = food.Consume(eaten, {eaten});

    CHECK(consumed);
    CHECK(food.Items().size() == 1);
    bool stillThere = std::find(food.Items().begin(), food.Items().end(), eaten) != food.Items().end();
    CHECK_FALSE(stillThere);
}

TEST_CASE("consume ignores a cell with no food") {
    FoodManager food(3, 3, 1, /*seed=*/42);
    food.Reset({});
    CHECK_FALSE(food.Consume(Cell{2, 2}, {}));
}
